File: Design HashMap/HashMap.java
import java.util.LinkedList;
import java.util.List;

/*
key-----> (HashFunction)----> Index
key % size
size=10000
Collision ho skta hai
Toh Ham chaining krte hai
chain bahut bada nahi hona chahiye nahi toh TC O(n) Lag Jayega
Worst Case mein chaining mein TC- O(n)
Then LoanFactor Comes into the Picture



*/

import java.util.*;

class MyHashMap {

    private static class Pair {
        int key;
        int value;

        Pair(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private int SIZE = 10000;
    private List<LinkedList<Pair>> buckets;

    public MyHashMap() {
        buckets = new ArrayList<>(SIZE);
        for (int i = 0; i < SIZE; i++) {
            buckets.add(new LinkedList<>());
        }
    }

    private int getBucketIndex(int key) {
        return key % SIZE;
    }

    public void put(int key, int value) {
        int bucketIndex = getBucketIndex(key);
        LinkedList<Pair> chain = buckets.get(bucketIndex);

        for (Pair pair : chain) {
            if (pair.key == key) {
                pair.value = value;   // update
                return;
            }
        }
        chain.add(new Pair(key, value));
    }

    public int get(int key) {
        int bucketIndex = getBucketIndex(key);
        LinkedList<Pair> chain = buckets.get(bucketIndex);

        for (Pair pair : chain) {
            if (pair.key == key) {
                return pair.value;
            }
        }
        return -1;
    }

    public void remove(int key) {
        int bucketIndex = getBucketIndex(key);
        LinkedList<Pair> chain = buckets.get(bucketIndex);

        Iterator<Pair> iterator = chain.iterator();
        while (iterator.hasNext()) {
            if (iterator.next().key == key) {
                iterator.remove();
                return;
            }
        }
    }
}

File: Design HashMap/Hashmap.cpp
class MyHashMap {
public:
    vector<list<pair<int, int>>> buckets;
    int size=10000;
    MyHashMap() {
        buckets.resize(size);
    }

    void put(int key, int value) {
        int bucketNumber = key % size;
        auto &chain = buckets[bucketNumber];
        for(auto &it : chain) {
            if(it.first == key) {
                it.second=value;
                return;
            }
        }
        chain.emplace_back(key, value);
    }

    int get(int key) {
        int bucketNumber = key % size;
        auto &chain = buckets[bucketNumber];
        if(chain.size()==0) return -1;
        for(auto &it: chain) {
            if(it.first == key) return it.second;
        }
        return -1;
    }

    void remove(int key) {
        int bucketNumber = key % size;
        auto &chain = buckets[bucketNumber];
        for(auto it = chain.begin(); it != chain.end(); it++) {
            if(it->first == key) {
                chain.erase(it);
                return;
            }
        }
    }
};